### ISSUE TITLE
Reduce time spent on visibility check in indexes

### ISSUE THREAD

**Issue by jrolli** [cmu-db/noisepage#510]
### Background
@mbutrovich has identified that a large amount of time during the TPC-C benchmark is spent on visibility checks on tupleslots during index operations.  The current method for the visibility check involves reading both two different sections of the block (allocation bitmap and the version pointer) and then unconditionally traversing the version chain as you would for materialization.

### Proposed Solution
Steal the low order bits of the version pointer to store visibility hints.  Specifically, our undo records are 64-bit aligned which gives us three low order bits that do not currently carry information.  We could use these bits to track allocation, "insert in version chain", and deleted.  This would allow a visibility check to read atomically read the version pointer and current visibility state together, and in the common case (allocated, insert not in the version chain, and not deleted) return true.

#### Flag definitions:
- **Unallocated**: Indicates whether the spot is unallocated (1) or allocated (0).  This semantic makes the visibility check cleaner.
- **Insert in chain**: Indicates whether the existing version chain ends in an insert (1) or the insert is no longer visible (0) [definition means this flag is cleared when it is safe to unlink but before that has happened].
- **Deleted**:  Indicates whether this tuple is logically deleted (1) or still exists (0).  A deletion will set this when it first stages the delete.

#### SqlTable behavior
- On initial allocation of a block, this solution would require scanning down the version pointer column and initializing it to set the unallocated (incurs some cost on inserts, existing allocated bitmap would be removed).
- On an insert, the atomic swap of the undo record would include clearing the "unallocated" bit and setting the "insert in chain" bit.
- On unlink of an insert record, the GC would first clear the "insert in chain" bit through repeated CAS until successful and then proceed with the normal unlink process.
- On a delete, the atomic swap adding the delete record would include setting the "deleted" bit.
- On unlink of a deletion, the GC would ensure the deleted bit remains set and issue a deferred action to set the "unallocated" bit.
- Other operations would remain generally unmodified other than requiring the lowest three bits to be masked out before dereferencing the pointer.

#### Index Visibility Check
1. Read the version information: if low order three bits equals 0, return true.  (Should generally be the common case)
2. Apply mask for the version pointer:  if the pointer is nullptr, return false. (either unallocated or deleted)
3. Chase the version chain.  This can potentially be optimized for early exiting, if needed.

**Comment by mbutrovich** [comment]
@LiuXiaoxuanPKU and I looked at this a bit over the weekend, trying to define baseline benchmarks and justification for this effort. As we investigated further, it appears the time being spent in `InVisible` by the index wrapper is actually due to repeated calls to the function, not necessarily because the function is slow. In high concurrency scenarios of TPC-C in particular, it seems like the deferred event framework is falling behind and not deleting keys from the index fast enough, so the wrapper ends up iterating through deleted keys calling `IsVisible` before it finds the key it wants.

Long story short: we need to profile the deferred event framework on a TPC-C workload that scales the number of workers.

**Comment by mbutrovich** [comment]
I should also add that it seems like #527 is a key contributor to this issue.
